euphonic/design/03_design_document_v0.3.md

```diff
@@ -0,0 +1,531 @@
+resolved per atom and quantum order. For Euphonic to be useful to Abins, the
+requirements are:
+
+- Euphonic must be able to produce a 1D energy-binned incoherent spectrum
+resolved per atom and quantum order
+- Euphonic must run with versions of Python that are shipped with Mantid (as of
+Mantid 4.3.0 Python 3.6, 3.7, 3.8)
+
+[1] P. C. H. Mitchell, S. F. Parker, A. J. Ramirez-Cuesta, J. Tomkinson, *Vibrational Spectroscopy With Neutrons*, World Scientific Publishing Co. Pte. Ltd., Singapore, 2005, 559-562
+ 
+## 3. Standalone (via Python API)
+
+Being able to efficiently calculate phonon frequencies and structure factors at
+arbitrary q-points is potentially useful beyond just the users of Horace and
+Abins e.g. calculating density of states, doing simple plotting from CASTEP's
+.phonon files. Because of this, there are several extra requirements that will
+enable Euphonic to be a useful standalone package:
+
+- The Python API should be clear and consistent (even if this means being
+slightly more verbose)
+- Functions that produce new data should return immutable, compartmentalised
+data objects rather than modifying the input object
+- Units should be handled carefully and be explicit
+- We should provide simple plotting functions
+- We should provide functions to serialise data objects (e.g to_dict, to_yaml)
+- We should provide functions to create data objects from these serialised
+formats (e.g. from_dict, from_yaml)
+
+## 4. For consideration: CASTEP dispersion.pl/dos.pl replacements
+
+Although not a requirement of PACE, some of Euphonic was initially developed to
+replace CASTEP's dispersion.pl/dos.pl scripts, which can plot a simple
+dispersion or DOS from CASTEP .phonon or .bands files. This is why the BandsData
+object and Grace plotting functions exist in
+[v0.2.2](01_design_overview_v0.2.2.md). 
+
+It has been decided this functionality will move to a separate repository,
+maintained and distributed from the CASTEP side of things. That repository can
+be found [here](https://bitbucket.org/castep/dispersiondos/src/master/).
+Having helper scripts to allow plotting and simple interpolation from the
+command line is also useful to Euphonic, so scripts could be developed within
+Euphonic, but in such a way that allows them to be wrapped so the extra .bands
+reading and Grace plotting functionality can be easily added without code
+duplication.
+
+Maintaining compatibility with CASTEP scripts and their requirements should be
+considered, but not a priority that should restrict Euphonic development, as the
+above scripts may not require new Euphonic features and could always depend on a
+specific version.
+
+## Context Diagram
+
+<img align="center" src="images/03_context.png">
+
+# Current Solution and Shortcomings
+
+See [v0.2.2 overview](01_design_overview_v0.2.2.md)
+
+# Proposed Solution
+
+## Euphonic Data Flow
+
+<img align="center" src="images/03_data_flow.png">
+
```

Comment:
I don't think we've decided on the Euphonic/Brille/Horace hierarchy yet, I think we shouldn't require Brille (as far as I know it has some relatively hefty dependencies), but we shouldn't preclude using it either. @g5t do you have any thoughts?

Reply:
Building the Python interface for `brille` requires:

1. a C++14 compliant compiler
2. CMake &ge; v3.13
3. pybind11 &ge; v2.2.0
4. [*optional but strongly recommended*] OpenMP &ge; v2.0

The first _shouldn't_ be difficult, but some linux systems have ancient system compilers (e.g., CentOS 7.7 uses gcc 4.8.5 which only has _experimental_ support for __*C++11*__).
To get around pre-C++-14 system compilers I was planning to build and distribute binaries on PyPI using [manylinux](github.com/pypa/manylinux). Thus far I haven't managed to build `brille` for `manylinux1` and the `manylinux2010` wheels I've made were ignored by `pip`.

At present CMake, pybind11 and OpenMP need to be installed system-wide, but it may be possible to reconfigure `brille`'s `setup.py` to use CMake and pybind11 from PyPI. If OpenMP isn't present `brille` should still compile and run.

If I sort out the manylinux issue and/or change the build to not rely on pre-installed system utilities/libraries then it would be simple to add `brille` as a dependency of `Euphonic`.

As to which hierarchy of

- `Horace` &rightarrow; `Euphonic` &rightarrow; `brille`
- `Horace` &rightarrow; `brille` &rightarrow; `Euphonic`

is desirable: ¯\\\_(ツ)\_/¯

Reply:
Ah, that's easier than I thought. I'll add some arrows to the context diagram and a mention this possibility in the Horace/Brille section. I can't think of anything in the design which would prevent us putting in an extra step to use Brille in the future.